celery/concurrency/asynpool.py

```diff
@@ -978,10 +978,14 @@ def _write_ack(fd, ack, callback=None):
     def flush(self):
         if self._state == TERMINATE:
             return
-        # cancel all tasks that haven't been accepted so that NACK is sent.
-        for job in self._cache.values():
+        # cancel all tasks that haven't been accepted so that NACK is sent
+        # if synack is enabled.
+        for job in tuple(self._cache.values()):
             if not job._accepted:
-                job._cancel()
+                if self.synack:
+                    job._cancel()
+                else:
+                    job.discard()
```

Comment:
FWICT `billiard.pool.Pool._cancel()` sets `_cancelled = True` and the it's up to (an eventual?) call to `Pool._ack()` to deal with that and send the NACK. What it's as clear to me is once the NACK is sent and `Pool._ack()` returns early at this line: https://github.com/celery/billiard/blob/9e89952f53cf8bd23fc115069b6fa1414fb94f44/billiard/pool.py#L1836 (or the line below if there is no `synqW_fd` passed in), does the job get discarded automagically? Or do we also need to cause a call to `job.discard()` to occur once that NACK has been sent?

In any case, this seems no worse than the current code which only cancels so LGTM

Reply:
It does seem we NACK based on the response.
https://github.com/celery/celery/blob/f462a437e3371acb867e94b52c2595b6d0a742d8/celery/concurrency/asynpool.py#L927-L937

What happens next is still beyond my grasp.